Fix #13249: Fix mypy pre-push checks

hardikkat24 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #13429 
2. This PR does the following: 
   - Adds the path to `USER_BASE/bin` to `_PATH_TO_INSERT` if mypy_requirements are installed using --user flag.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes made in this PR work correctly.
If the changes in your PRs are autogenerated via a script and you cannot provide proof for the changes then please leave a comment "No proof of changes needed because {{Reason}}".
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

hardikkat24 committed:
fix

hardikkat24 committed:
workflow

hardikkat24 committed:
workflow added

hardikkat24 commented:
@EricZLou @EeshaArif Please test this PR.
Steps:
1. Pull from this PR
2. Remove third_party/python3_libs.
3. Run mypy script. 

Please let me know if you face any issues. Thanks!

EeshaArif commented:
Thanks @hardikkat24 still getting this, PTAL.
```
Installing Mypy and stubs for third party libraries.
Installed Mypy and stubs for third party libraries.
Starting Mypy type checks.
Traceback (most recent call last):
  File "/usr/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/eesha/Desktop/opensource/oppia/scripts/run_mypy_checks.py", line 897, in <module>
    main()
  File "/home/eesha/Desktop/opensource/oppia/scripts/run_mypy_checks.py", line 881, in main
    cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, env=env)
  File "/usr/lib/python2.7/subprocess.py", line 394, in __init__
    errread, errwrite)
  File "/usr/lib/python2.7/subprocess.py", line 1047, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory

```

hardikkat24 commented:
@EeshaArif can you please confirm whether the bin Directory is created in USER_BASE? And send the directory tree? Thanks!!

EeshaArif commented:
yes, it is the same as stated [here](https://github.com/oppia/oppia/issues/13429#issuecomment-884924717)

hardikkat24 commented:
@EeshaArif thank you. Will try one more thing and get back to you. Thanks!!

hardikkat24 commented:
@DubeySandeep @vojtechjelinek @sajalasati Tagging you since I'm unable to fix this issue. 
This is the same issue we fixed earlier for mac systems. Now we have changed the global installation to installation in a particular directory. (third_party/python3_libs)
Now in some systems in which normal pip installation throws error, we install them using by adding more arguments: --user, --prefix=, --system . You can see [here](https://github.com/oppia/oppia/blob/8804c64fc08289f15fb8b3f95a0ce3bf67db3091/scripts/run_mypy_checks.py#L833). In these systems, the error  'No such file or directory' is thrown continuously. Thanks!!

hardikkat24 committed:
changes

hardikkat24 committed:
global introduced again

hardikkat24 commented:
@EricZLou @EeshaArif Can you please test this PR again? Thanks!!

hardikkat24 commented:
@EeshaArif @EricZLou Please test this. Thanks!!

EeshaArif commented:
Tested it, its passing... Thanks @hardikkat24

hardikkat24 commented:
@DubeySandeep @seanlip PTAL.

seanlip reviewed (approved):
LGTM -- thanks @hardikkat24. No issues from my side as long as it works on other people's machines.

Just one question -- why don't we always install locally (never globally)?

U8NWXD commented:
Removing review request for @U8NWXD and @krishnarao22 since we are not code owners. If you want us to review anyway, feel free to add us back!

hardikkat24 commented:
@seanlip This is a workaround we figured out to run CI checks. If we were not installing mypy globally, mypy was not detected on CI and hence the checks were not running. We already have an [issue](https://github.com/oppia/oppia/issues/13398) for this and will be taking a look at this when py3 migration is complete. Thanks!

seanlip commented:
Sounds good. Thanks for the explanation, @hardikkat24! I'll merge this PR.
